include/nbl/asset/utils/CVertexWelder.h

```diff
@@ -0,0 +1,237 @@
+            return (1.0 - hlsl::dot(val1, val2)) < epsilon;
+          }
+        }
+        return true;
+      }
+
+      float m_epsilon;
+
+    public:
+
+      DefaultWeldPredicate(float epsilon) : m_epsilon(epsilon) {}
+
+      bool operator()(const ICPUPolygonGeometry* polygon, uint32_t index1, uint32_t index2)
+      {
+        if (!isAttributeValEqual(polygon->getPositionView(), index1, index2, m_epsilon))
+          return false;
+        if (!isAttributeDirEqual(polygon->getNormalView(), index1, index2, m_epsilon))
+          return false;
+        for (const auto& jointWeightView : polygon->getJointWeightViews())
+        {
+          if (!isAttributeValEqual(jointWeightView.indices, index1, index2, m_epsilon)) return false;
+          if (!isAttributeValEqual(jointWeightView.weights, index1, index2, m_epsilon)) return false;
+        }
+        for (const auto& auxAttributeView : polygon->getAuxAttributeViews())
+          if (!isAttributeValEqual(auxAttributeView, index1, index2, m_epsilon)) return false;
+
+        return true;
+      }
+        
+  };
+
+  template <typename AccelStructureT>
+  static core::smart_refctd_ptr<ICPUPolygonGeometry> weldVertices(const ICPUPolygonGeometry* polygon, const AccelStructureT& as, WeldPredicateFn shouldWeldFn) {
+    auto outPolygon = core::move_and_static_cast<ICPUPolygonGeometry>(polygon->clone(0u));
+    outPolygon->setIndexing(IPolygonGeometryBase::TriangleList());
+
+    core::vector<uint32_t> vertexIndexToAsIndex(as.getVertexCount());
+
+    for (uint32_t vertexData_i = 0u; vertexData_i < as.getVertexCount(); vertexData_i++)
+    {
+      const auto& vertexData = as.vertices()[vertexData_i];
+      vertexIndexToAsIndex[vertexData.index] = vertexData.index;
+    }
+
+    static constexpr auto INVALID_INDEX = std::numeric_limits<uint32_t>::max();
+    core::vector<uint32_t> remappedVertexIndexes(as.getVertexCount());
+    std::fill(remappedVertexIndexes.begin(), remappedVertexIndexes.end(), INVALID_INDEX);
+
+    uint32_t maxRemappedIndex = 0;
+    // iterate by index, so that we always use the smallest index when multiple vertexes can be welded together
+    for (uint32_t index = 0; index < as.getVertexCount(); index++)
+    {
+      const auto asIndex = vertexIndexToAsIndex[index];
+      const auto& vertexData = as.vertices()[asIndex];
+      auto& remappedVertexIndex = remappedVertexIndexes[index];
+      as.forEachBroadphaseNeighborCandidates(vertexData, [&, polygon, index](const typename AccelStructureT::vertex_data_t& neighbor) {
+        const auto neighborRemappedIndex = remappedVertexIndexes[neighbor.index];
+        if (shouldWeldFn(polygon, index, neighbor.index) && neighborRemappedIndex != INVALID_INDEX) {
+          remappedVertexIndex = neighborRemappedIndex;
+          return false;
+        }
+        return true;
+      });
+      if (remappedVertexIndex != INVALID_INDEX) {
+        remappedVertexIndex = vertexData.index;
+        maxRemappedIndex = vertexData.index;
```

Comment:
no, not the max of the original vertex index, you want the max you FOUND so `if (maxRemappedIndex<remappedVertexIndex) maxRemappedIndex = remappedVertexIndex`